content/200-concepts/100-components/01-prisma-schema/06-relations/07-referential-actions.mdx

```diff
@@ -8,6 +8,401 @@ preview: true
+
+</Admonition>
+
+### NoAction
+
+- `onDelete: NoAction` Prevents the deletion if any referencing records exist.
+
+- `onUpdate: NoAction` Prevents the identifier of a referenced record from being changed.
+
+The `NoAction` action is similar to `Restrict` in that it will prevent the update or deletion of any records which are connected and throw an error. The difference between the two is dependant on the database being used.
+
+#### Example usage
+
+```prisma file=schema.prisma highlight=4;add
+model Post {
+  id        Int     @id @default(autoincrement())
+  title     String
+  author    User    @relation(fields: [authorId], references: [id], onDelete: NoAction, onUpdate: NoAction)
+  authorId  Int
+}
+
+model User {
+  id    Int    @id @default(autoincrement())
+  posts Post[]
+}
+```
+
+##### Result of using `NoAction`?
+
+`User`'s with posts **cannot** be deleted. The `Post`'s `id` **cannot** be changed.
+
+### SetNull
+
+- `onDelete: SetNull` The scalar field of the referenced object will be set to `NULL`.
+
+- `onUpdate: SetNull` When updating the identifier of a referenced object, the scalar fields of the referencing objects will be set to `NULL`.
+
+If one or more of the relations scalar fields are required a runtime error will be thrown.
+
+```prisma file=schema.prisma highlight=4;add
+model Post {
+  id        Int     @id @default(autoincrement())
+  title     String
+  author    User    @relation(fields: [authorId], references: [id], onDelete: SetNull, onUpdate: SetNull)
+  authorId  Int
+}
+
+model User {
+  id    Int    @id @default(autoincrement())
+  posts Post[]
+}
+```
+
+##### Result of using `SetNull`?
+
+When deleting a `User`, the `authorId` will be set to `NULL` for all its authored posts. 
+
+When changing a `User`'s `id`, the `authorId` will be set to `NULL` for all its authored posts. 
+
+### SetDefault
+
+- `onDelete: SetDefault` The scalar field of the referencing object will be set to the fields default value.
+
+- `onUpdate: SetDefault` The scalar field of the referencing object will be set to the fields default value.
+
+These require setting a default for the relation scalar field with [`@default`](../../../../reference/api-reference/prisma-schema-reference#default). If no defaults are provided for any of the scalar fields, a runtime error will be thrown.
+
+```prisma file=schema.prisma highlight=4,5;add
+model Post {
+  id          Int     @id @default(autoincrement())
+  title       String
+  authorName  String  @default("anonymous")
+  author      User    @relation(fields: [authorId], references: [id], onDelete: SetDefault, onUpdate: SetDefault)
+  authorId    Int
+}
+
+model User {
+  id        Int    @id @default(autoincrement())
+  username  String @id
+  posts     Post[]
+}
```

Comment:
There is something weird in this example. Maybe it was an oversight in my original example if you copied and pasted it?

* There are two scalar fields that look like relation scalar fields (authorName and authorId). 
* The relation is using authorId so the columns authorName and username would not be needed
* If using these fields, then we would need to a default value to the authorId column. I avoided using this in my original example because having a default value for an integer based ID seems wrong.


```suggestion
model Post {
  id          Int     @id @default(autoincrement())
  title       String
  authorName  String  @default("anonymous")
  author      User    @relation(fields: [authorName], references: [id], onDelete: SetDefault, onUpdate: SetDefault)
}

model User {
  username  String @id
  posts     Post[]
}
```

We should probably add a note that this assumes a record exists in the table User with username="anonymous", otherwise referential integrity would be violated and the action would fail (delete or update prevented).

Reply:
Alternatively, we could the username and authorName columns, and add `@default: 1` to the authorId column. I like this option less.